backend/dataPreprocessor.js

```diff
@@ -0,0 +1,93 @@
+const { PrismaClient } = require("@prisma/client");
+const prisma = new PrismaClient();
+const fs = require("fs");
+const natural = require("natural");
+const { TfIdf } = natural;
+const tfidf = new TfIdf();
+const tf = require("@tensorflow/tfjs-node");
+
+// Fetch data from Prisma
+const fetchData = async () => {
+  return await prisma.interaction.findMany({
+    include: {
+      User: true,
+      Business: true,
+    },
+  });
+};
+
+// Handle missing values
+const handleMissingValues = (data) => {
+  return data.map((item) => {
+    item.rated = item.rated !== null ? item.rated : 0;
+    item.reviewed = item.reviewed !== null ? item.reviewed : "";
+    return item;
+  });
+};
+
+// Normalize text data
+const normalizeTextData = (data) => {
+  return data.map((item) => {
+    item.reviewed = item.reviewed.toLowerCase().replace(/[^a-z0-9\s]/g, "");
```

Comment:
Most people don't know regex; I assume this is sourced from somewhere, can you add what the query does in the comments for visibilty? Looks like it checks for alphanumerics.